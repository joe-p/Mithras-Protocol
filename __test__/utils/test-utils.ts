import circom_tester from "circom_tester";
import { AlgorandClient } from "@algorandfoundation/algokit-utils/types/algorand-client";
import { microAlgos, Config } from "@algorandfoundation/algokit-utils";
import algosdk from "algosdk";

Config.configure({
  logger: {
    error: () => {},
    warn: () => {},
    info: () => {},
    verbose: () => {},
    debug: () => {},
  },
});

const wasm_tester = circom_tester.wasm;

export interface CircuitTestConfig {
  circuitPath: string;
  prime?: string;
  recompile?: boolean;
}

export interface MimcInput {
  msgs: [bigint, bigint];
}

export interface MerklePathInput {
  leaf: bigint;
  pathElements: bigint[];
  pathSelectors: number[];
}

export class CircuitTester {
  static async create(config: CircuitTestConfig) {
    return await wasm_tester(config.circuitPath, {
      prime: config.prime || "bls12381",
      recompile: config.recompile ?? true,
    });
  }

  static async createMimcTester() {
    return await this.create({
      circuitPath: "__test__/circuits/mimc.circom",
    });
  }

  static async createMerklePathTester() {
    return await this.create({
      circuitPath: "__test__/circuits/merkle_path_verify.circom",
    });
  }

  static async createMimcSum4Tester() {
    return await this.create({
      circuitPath: "__test__/circuits/mimc_sum4.circom",
    });
  }
}

export class AlgorandTestUtils {
  static createLocalClient(): AlgorandClient {
    return AlgorandClient.defaultLocalNet();
  }

  static async getDispenser(client: AlgorandClient) {
    return await client.account.localNetDispenser();
  }

  static generateRandomAppName(prefix: string): string {
    return `${prefix}-${Math.random().toString(36).substring(2, 15)}`;
  }

  static async fundAccount(
    client: AlgorandClient,
    address: algosdk.Address,
    amount: number,
  ) {
    await client.account.ensureFundedFromEnvironment(
      address,
      microAlgos(amount),
    );
  }
}

export class MimcCalculator {
  private circuit: any;
  private sum4?: any;

  constructor(circuit: any, sum4?: any) {
    this.circuit = circuit;
    this.sum4 = sum4;
  }

  static async create(): Promise<MimcCalculator> {
    const circuit = await CircuitTester.createMimcTester();
    const sum4 = await CircuitTester.createMimcSum4Tester();
    return new MimcCalculator(circuit, sum4);
  }

  async calculateHash(left: bigint, right: bigint): Promise<bigint> {
    const witness = await this.circuit.calculateWitness({
      msgs: [left, right],
    });
    return witness[1];
  }

  async sum4Commit(msgs: [bigint, bigint, bigint, bigint]): Promise<bigint> {
    if (!this.sum4) throw new Error("sum4 circuit not initialized");
    const witness = await this.sum4.calculateWitness({ msgs });
    return witness[1];
  }

  async calculateMerkleRoot(
    leaf: bigint,
    pathElements: bigint[],
    pathSelectors: number[],
  ): Promise<bigint> {
    let currentHash = leaf;

    for (let i = 0; i < pathElements.length; i++) {
      const left = pathSelectors[i] === 0 ? currentHash : pathElements[i];
      const right = pathSelectors[i] === 0 ? pathElements[i] : currentHash;

      currentHash = await this.calculateHash(left, right);
    }

    return currentHash;
  }
}

export class MerkleTestHelpers {
  static createDefaultPathElements(size: number = 32): bigint[] {
    return new Array(size).fill(0n);
  }

  static createDefaultPathSelectors(size: number = 32): number[] {
    return new Array(size).fill(0);
  }

  static bytesToBigInt(bytes: Uint8Array): bigint {
    return BigInt("0x" + Buffer.from(bytes).toString("hex"));
  }

  static bigIntToBytes(value: bigint): Uint8Array {
    const hex = value.toString(16).padStart(64, "0");
    const bytes = new Uint8Array(32);
    for (let i = 0; i < 32; i++) {
      bytes[i] = parseInt(hex.substr(i * 2, 2), 16);
    }
    return bytes;
  }

  static async verifyCircuitWithInput(
    circuit: any,
    input: MerklePathInput,
  ): Promise<void> {
    const witness = await circuit.calculateWitness(input);
    await circuit.checkConstraints(witness);
  }
}

export class TestDataBuilder {
  static createMimcInput(left: bigint, right: bigint): MimcInput {
    return { msgs: [left, right] };
  }

  static createMerklePathInput(
    leaf: bigint,
    pathElements: bigint[] = MerkleTestHelpers.createDefaultPathElements(),
    pathSelectors: number[] = MerkleTestHelpers.createDefaultPathSelectors(),
  ): MerklePathInput {
    return { leaf, pathElements, pathSelectors };
  }

  static createTestLeaf(value: number): Uint8Array {
    const leaf = new Uint8Array(32);
    const bytes = [
      (value >> 24) & 0xff,
      (value >> 16) & 0xff,
      (value >> 8) & 0xff,
      value & 0xff,
    ];
    leaf.set(bytes, 28);
    return leaf;
  }
}
