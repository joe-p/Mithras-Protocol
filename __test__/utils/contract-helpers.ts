import { microAlgos } from "@algorandfoundation/algokit-utils";
import {
  MimcMerkleFactory,
  MimcMerkleClient,
} from "../../contracts/clients/MimcMerkle";
import { MimcTestClient, MimcTestFactory } from "../contracts/clients/MimcTest";
import { AlgorandTestUtils } from "./test-utils";

export class MimcTestHelper {
  static async deployContract(): Promise<MimcTestClient> {
    const algorand = AlgorandTestUtils.createLocalClient();
    const factory = new MimcTestFactory({
      algorand,
      defaultSender: await AlgorandTestUtils.getDispenser(algorand),
    });

    const { appClient } = await factory.deploy({
      appName: AlgorandTestUtils.generateRandomAppName("mimc-test"),
    });

    return appClient;
  }
}

export class MimcMerkleHelper {
  static async deployContract(): Promise<MimcMerkleClient> {
    const algorand = AlgorandTestUtils.createLocalClient();
    const factory = new MimcMerkleFactory({
      algorand,
      defaultSender: await AlgorandTestUtils.getDispenser(algorand),
    });

    const { appClient } = await factory.deploy({
      appName: AlgorandTestUtils.generateRandomAppName("mimc-merkle"),
    });

    await AlgorandTestUtils.fundAccount(
      algorand,
      appClient.appAddress,
      1567900,
    );

    await appClient.send.bootstrap({
      args: {},
      extraFee: microAlgos(256 * 1000),
    });

    return appClient;
  }

  static async addLeaf(
    appClient: MimcMerkleClient,
    leafHash: Uint8Array,
  ): Promise<void> {
    await appClient.send.addLeaf({
      args: { leafHash },
      extraFee: microAlgos(256 * 1000),
    });
  }

  static async getContractState(appClient: MimcMerkleClient) {
    const subtree = await appClient.state.box.subtree();
    const zeroHashes = await appClient.state.box.zeroHashes();

    if (!subtree || !zeroHashes) {
      throw new Error("Failed to get state from contract");
    }

    return { subtree, zeroHashes };
  }

  static async verifyRoot(
    appClient: MimcMerkleClient,
    root: Uint8Array,
  ): Promise<boolean> {
    const { return: isValid } = await appClient.send.isValidRoot({
      args: { root },
    });
    return isValid!;
  }
}

