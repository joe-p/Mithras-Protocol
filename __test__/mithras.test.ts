import { AlgorandClient, microAlgos } from "@algorandfoundation/algokit-utils";
import {
  AppVerifier,
  LsigVerifier,
  SignalsAndProofClient,
  SignalsAndProofFactory,
} from "snarkjs-algorand";
import { MithrasClient, MithrasFactory } from "../contracts/clients/Mithras";

import { beforeAll, describe, expect, it } from "vitest";
import { MerkleTestHelpers } from "./utils/test-utils";

const DEPOSIT_LSIGS = 6;
const SPEND_LSIGS = 14;

describe("Mithras App", () => {
  let depositVerifier: LsigVerifier;
  let spendVerifier: LsigVerifier;
  let appClient: MithrasClient;
  let algorand: AlgorandClient;
  let signalsAndProofClient: SignalsAndProofClient;

  beforeAll(async () => {
    algorand = AlgorandClient.defaultLocalNet();
    algorand.setSuggestedParamsCacheTimeout(0);
    depositVerifier = new LsigVerifier(
      algorand,
      "circuits/deposit_test.zkey",
      "circuits/deposit_js/deposit.wasm",
      DEPOSIT_LSIGS,
    );

    spendVerifier = new LsigVerifier(
      algorand,
      "circuits/spend_test.zkey",
      "circuits/spend_js/spend.wasm",
      SPEND_LSIGS,
    );

    const signalsAndProofFactory = new SignalsAndProofFactory({
      defaultSender: await algorand.account.localNetDispenser(),
      algorand,
    });

    const { appClient: sapc } = await signalsAndProofFactory.deploy();

    signalsAndProofClient = sapc;

    const factory = new MithrasFactory({
      algorand,
      defaultSender: await algorand.account.localNetDispenser(),
    });

    const { appClient: ac } = await factory.send.create.createApplication({
      args: {
        depositVerifier: (await depositVerifier.lsigAccount()).addr.toString(),
        spendVerifier: (await spendVerifier.lsigAccount()).addr.toString(),
      },
    });

    appClient = ac;

    // TODO: determine the actual MBR needed
    await appClient.appClient.fundAppAccount({ amount: microAlgos(4848000) });

    await appClient.send.bootstrapMerkleTree({
      args: {},
      // TODO: determine the actual fee needed
      extraFee: microAlgos(256 * 1000),
    });
  });

  it("deposit", async () => {
    const spending_secret = 1n;
    const nullifier_secret = 2n;
    const amount = 3n;
    const receiver = 4n;

    const group = appClient.newGroup();

    await depositVerifier.verify({
      inputs: {
        spending_secret,
        nullifier_secret,
        amount,
        receiver,
      },
      callback: async (arg) => {
        const { appParams, extraLsigsTxns } = arg;

        // App call from lsig to expose the signals and proof to our app
        const signalsAndProofCall = (
          await signalsAndProofClient.createTransaction.signalsAndProof(
            appParams,
          )
        ).transactions[0];

        group.deposit({
          args: { signalsAndProofCall },
          extraFee: microAlgos(256 * 1000),
        });

        for (const txn of extraLsigsTxns) {
          group.addTransaction(txn);
        }
      },
    });

    const simRes = await group.simulate({
      allowUnnamedResources: true,
    });

    expect(
      simRes.simulateResponse.txnGroups[0].appBudgetConsumed,
    ).toMatchSnapshot("deposit app budget");
  });

  it("spend", async () => {
    const utxo_spending_secret = 11n;
    const utxo_nullifier_secret = 22n;
    const utxo_amount = 33n;
    const utxo_spender = 44n;

    const depositGroup = appClient.newGroup();

    await depositVerifier.verify({
      inputs: {
        spending_secret: utxo_spending_secret,
        nullifier_secret: utxo_nullifier_secret,
        amount: utxo_amount,
        receiver: utxo_spender,
      },
      callback: async (arg) => {
        const { appParams, extraLsigsTxns } = arg;

        // App call from lsig to expose the signals and proof to our app
        const signalsAndProofCall = (
          await signalsAndProofClient.createTransaction.signalsAndProof(
            appParams,
          )
        ).transactions[0];

        depositGroup.deposit({
          args: { signalsAndProofCall },
          extraFee: microAlgos(256 * 1000),
        });

        for (const txn of extraLsigsTxns) {
          depositGroup.addTransaction(txn);
        }
      },
    });

    await depositGroup.send();

    const fee = 0n;
    const out0_amount = 11n;
    const out1_amount = 22n;
    const out0_receiver = 1234n;
    const out1_receiver = 5678n;
    const out0_spending_secret = 333n;
    const out0_nullifier_secret = 444n;
    const out1_spending_secret = 555n;
    const out1_nullifier_secret = 666n;

    // Build a trivial path: all zeros to compute root
    const pathElements = MerkleTestHelpers.createDefaultPathElements();
    const pathSelectors = MerkleTestHelpers.createDefaultPathSelectors();

    const inputSignals = {
      fee,
      utxo_spender,
      utxo_spending_secret,
      utxo_nullifier_secret,
      utxo_amount,
      path_selectors: pathSelectors,
      utxo_path: pathElements,
      out0_amount,
      out0_receiver,
      out0_spending_secret,
      out0_nullifier_secret,
      out1_amount,
      out1_receiver,
      out1_spending_secret,
      out1_nullifier_secret,
    };

    const spendGroup = appClient.newGroup();

    await spendVerifier.verify({
      inputs: inputSignals,
      callback: async (arg) => {
        const { appParams, extraLsigsTxns } = arg;

        // App call from lsig to expose the signals and proof to our app
        const signalsAndProofCall = (
          await signalsAndProofClient.createTransaction.signalsAndProof(
            appParams,
          )
        ).transactions[0];

        spendGroup.spend({
          args: { verifierCall: signalsAndProofCall },
          extraFee: microAlgos(256 * 1000),
        });

        for (const txn of extraLsigsTxns) {
          spendGroup.addTransaction(txn);
        }
      },
    });

    const simRes = await spendGroup.simulate({
      allowUnnamedResources: true,
    });

    expect(
      simRes.simulateResponse.txnGroups[0].appBudgetConsumed,
    ).toMatchSnapshot("spend app budget");
  });
});
