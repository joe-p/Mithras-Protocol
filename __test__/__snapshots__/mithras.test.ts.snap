// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Mithras App > deposit > deposit app budget 1`] = `174391`;

exports[`Mithras App > spend > deposit app budget 1`] = `356661`;
